sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/TraceHelper.cs

```diff
@@ -300,5 +300,102 @@ private static void AddTelemetryFromActivityEvents(Activity activity, TelemetryI
+                {
+                    measurements["timeSinceEnqueued"] = enqueuedTime;
+                }
+            }
+        }
+
+        private static bool TryGetAverageTimeInQueueForBatchAndAddLinksToProperties(Activity activity, ref AzMonList UnMappedTags, out long avgTimeInQueue)
+        {
+            avgTimeInQueue = 0;
+            var linksCount = 0;
+            DateTimeOffset startTime = activity.StartTimeUtc;
+            long startEpochTime = startTime.ToUnixTimeMilliseconds();
+            bool isEnqueuedTimeCalculated = true;
+
+            string msLinks = "_MS.links";
+            var linksJson = new StringBuilder();
+            linksJson.Append('[');
+            foreach (ref readonly var link in activity.EnumerateLinks())
+            {
+                if (!TryGetEnqueuedTime(link, out var msgEnqueuedTime))
+                {
+                    // instrumentation does not consistently report enqueued time, ignoring whole span
+                    isEnqueuedTimeCalculated = false;
+                }
+
+                avgTimeInQueue += Math.Max(startEpochTime - msgEnqueuedTime, 0);
+
+                linksCount++;
+
+                if (linksCount <= MaxlinksAllowed)
+                {
+                    linksJson
+                        .Append('{')
+                        .Append("\"operation_Id\":")
+                        .Append('\"')
+                        .Append(link.Context.TraceId.ToHexString())
+                        .Append('\"')
+                        .Append(',');
+                    linksJson
+                        .Append("\"id\":")
+                        .Append('\"')
+                        .Append(link.Context.SpanId.ToHexString())
+                        .Append('\"');
+                    linksJson.Append("},");
+                }
+            }
+
+            if (linksJson.Length > 0)
+            {
+                // trim trailing comma - json does not support it
+                linksJson.Remove(linksJson.Length - 1, 1);
+            }
+
+            linksJson.Append(']');
+
+            AzMonList.Add(ref UnMappedTags, new KeyValuePair<string, object?>(msLinks, linksJson.ToString()));
+
+            if (MaxlinksAllowed < linksCount)
+            {
+                AzureMonitorExporterEventSource.Log.WriteInformational("ActivityLinksIgnored", $"Max count of {MaxlinksAllowed} has reached.");
+            }
+
+            if (isEnqueuedTimeCalculated)
+            {
+                avgTimeInQueue /= linksCount;
+            }
+            else
+            {
+                avgTimeInQueue = 0;
+                return false;
+            }
+
+            return true;
+        }
+
+        private static bool TryGetEnqueuedTime(ActivityLink link, out long enqueuedTime)
+        {
+            enqueuedTime = 0;
+            if (link.Tags != null && link.Tags.Any())
+            {
+                foreach (var attribute in link.Tags)
```

Comment:
Does `foreach (ref readonly var` work here? if so we could avoid the if check above.